### PR TITLE
1.x: unified onErrorX and onExceptionResumeNext and fixed backpressure

### PR DESCRIPTION
**Pull request by akarnokd**
The operators didn't honor backpressure when they switched over to the other Observable.

This PR also includes the fix #3685.

### CODE REVIEW

**Comment by akarnokd** [comment]
There was a test failure in BackpressureTests.testMergeAsync, run before any of the changes of this PR and shouldn't have any effect. Overloaded servers?

File: src/test/java/rx/internal/operators/OperatorOnErrorReturnTest.java [hunk 1]
```diff
@@ -217,6 +219,33 @@ public void run() {
...
+        
+        ps.onNext(1);
+        ps.onNext(2);
+        ps.onError(new TestException("Forced failure"));
+
+        ts.assertValues(1, 2);
+        ts.assertNoErrors();
+        ts.assertNotCompleted();
+
+        ts.requestMore(2);
```
Comment (stevegury) [inline src/test/java/rx/internal/operators/OperatorOnErrorReturnTest.java]: Not a big deal, but it should be `requestMore(1)` if you want to have the same behavior (than the other tests).

**Comment by stevegury** [comment]
:+1:

File: src/test/java/rx/internal/operators/OperatorOnErrorReturnTest.java (hunk 1 above)
Comment (akarnokd) [inline src/test/java/rx/internal/operators/OperatorOnErrorReturnTest.java]: It's deliberately overrequesting here.

**Pull request merged**